### PR TITLE
final project

### PR DESCRIPTION
**Pull request by sheylaPozo**
### Gems:

- Using rspec-rails gem for testing the models.
- Using shoulda_matchers gem for easier testing the associations and validations in the models.
- Using devise gem as authentication solution.
- Using cancancan gem as authorization solution.

### Database:

- Postgresql.

### Create models:
**User:**

**Associations:**

- A user has many groups and budgets.

### Validations:

- Name and email should be present, with a max length of 50 chars.
- Password should be present, with a max length of 30 chars.

### Group:

**Associations:**
A group belongs to the user and has many budgets.
Validations:
The name should be present, with a max length of 50 chars.
### Associations:

- A group belongs to the user and has many groups.

### Validations:

- The name should be present, with a max length of 50 chars.
- The amount should be present, be an integer, and be greater than 0.

### Create controllers:

- User.
- Group.
- Budget.
- Tests:
- Models.
- Controllers/requests.
- Views.

[live link](https://budgetapps.herokuapp.com/)

Thank you for your kind review! 😃❄

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -8,9 +8,6 @@
   root "groups#index", as: "budget"
```
Comment (rloterh) [inline config/routes.rb]: - [x] An **error is thrown** in an attempt to create a category when the _create_ button is pressed. Kindly ensure to debug and improve upon this aspect.

**Error:**
image

File: spec/features/budget/new.html.erb_spec.rb [hunk 2]
```diff
@@ -2,7 +2,7 @@
```
Comment (rloterh) [inline spec/features/budget/new.html.erb_spec.rb]: - [x] There are some test specs **failing**. Kindly ensure to improve upon those aspects.

**Failing rspec tests:**
image

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Shey, 

You did great on this project, and for making the changes requested in the previous review(s)! 👏👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦  🚦  

## Good Points 👍
- Descriptive Pull Request
- Good video presentaton
- Awesome ReadMe
- No linter errors

## Required Changes ♻️
_Kindly check the comments under the review._

<br>
Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 4 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: config/routes.rb (hunk 1 above)
Comment (sheylaPozo) [reply config/routes.rb]: corrected it

File: spec/features/budget/new.html.erb_spec.rb (hunk 2 above)
Comment (sheylaPozo) [reply spec/features/budget/new.html.erb_spec.rb]: image

Comment (sheylaPozo) [reply spec/features/budget/new.html.erb_spec.rb]: corrected

File: spec/factories/budgets.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
```
Comment (Davidosky007) [inline spec/factories/budgets.rb]: * Kindly fix this error i get when i try to run test 
image

File: app/views/categories/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+<%= form_with url: categories_path, model: @group do |f| %>
+  <header class="header">
```
Comment (Davidosky007) [inline app/views/categories/new.html.erb:2]: * As the previous reviewer pointed out that an error that occurs when trying to save a category still persists Kindly fix it 🙏 
image

**Review by Davidosky007 — CHANGES_REQUESTED** [review]
## Hi @sheylaPozo 👋, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/categories/new.html.erb (hunk 4 above)
Comment (sheylaPozo) [reply app/views/categories/new.html.erb:2]: image
fixed it

File: spec/factories/budgets.rb (hunk 3 above)
Comment (sheylaPozo) [reply spec/factories/budgets.rb]: fixed it

Comment (sheylaPozo) [reply spec/factories/budgets.rb]: apps

File: app/views/categories/new.html.erb (hunk 4 above)
Comment (sheylaPozo) [reply app/views/categories/new.html.erb:2]: fixed it
desktop

**Comment by Davidosky007** [comment]
## Hi @sheylaPozo 👋, 

## Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
## Congratulations! 🎉 


job-welldone-gif

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by sheylaPozo** [comment]
thank you so much :)

**Pull request merged**